Add feedback popups

beck-thompson opened the pull request:
## About the PR
<!-- What did you change? -->
Added feedback popups that appear at the end of the round. Right now, these just link to the forms directing people where they should give feedback!

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
#38811 showed that a lot of people do not know about the forms and its very difficult for the average player to give feedback. This PR does a couple things:
1.) Makes it clear what PRs are still being tested out and are not finalized
2.) Make it very easy for any player to know where to give their feedback.

## Technical details
<!-- Summary of code changes for easier review. -->
System is pretty simple but my main issue was naming stuff. Sever sends messages to clients at the end of the round, clients listen for the messages and then display popups showing what PRs need feedback.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

https://github.com/user-attachments/assets/b29cc177-f5f7-4cd4-bb23-6e0c3f7a4d29

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl: Beck
- add: Added feedback popups that players get at the end of the round! They will list all "test" PRs or PRs that people want feedback on. Right now they will link to a form thread

beck-thompson commented:
Based off of https://github.com/DeltaV-Station/Delta-v/pull/2561

Princess-Cheeseballs reviewed (changes requested):
Content.Client/DeltaV/?

VasilisThePikachu commented on Resources/Prototypes/FeedbackPopup/feedbackpopups.yml:
```diff
@@ -0,0 +1,8 @@
+  title: "Feedback on the new feedback popup system"
+  description:
+  - "This is a new system to get feedback on features. It will give popups at the end (and potentially middle) of the round!"
+  - "Please give any thoughts you have on this by going to the form below. You will use your ss14 username and password!"
```
Please revise, I guess forks can change this anyway but
- I would just say "login with your ss14 account"
- if you rather keep this wording, advice that login details shall only be entered on official sites and not be shared to anyone else.

beck-thompson commented:
What do you think of this?

https://github.com/user-attachments/assets/b4582819-fddd-4c03-a08c-ac1a003bb016

I added some indents to the text and added separators - I can also do something more like you posted if you want (it just requires a little more work :laughing:). Let me know!

SlamBamActionman commented:
> What do you think of this?

Looks good; I'll try it ingame tomorrow and with multiline just to get a feel myself and if I have nothing else to say I'll merge it :)

beck-thompson commented:
> Question regarding this: Is there a way for game admins to control what is shown, even if it's only for the lifetime of the current server process? As an example with the recent chem changes being proto-uploaded and overwritten.

Yes! You should just be able to proto upload a new feedback popup see:

<img width="930" height="163" alt="image" src="https://github.com/user-attachments/assets/c5dfc806-55ac-4524-85f3-501516745b03" />

I'm pretty sure you guys should be able to also change the CCVAR (E.g if you *only* want yours to show for a certain round or something) but I'm not 100% certain what permissions are needed / what perms admins have

SlamBamActionman commented:
Seperation looks good but it doesn't handle long text well. You need to make the button container's size be dominant and its height locked. 👍 

<img width="783" height="766" alt="image" src="https://github.com/user-attachments/assets/edfd3112-565b-44fc-ae8f-455459cfc513" />

beck-thompson commented:
UI is a little janky, if you have enough popups to where the UI needs a scroll bar it doesn't "refresh" and it looks like there are fewer things until you scroll. Just UI glitch, everything else works as intended

> If there is no ResponseLink listed, it will list a button. And error out when clicked.
Can't fix engine sandbox issue

perryprog commented on Content.Client/FeedbackPopup/FeedbackEntry.xaml.cs:
```diff
@@ -0,0 +1,55 @@
+            _uri.OpenUri(_feedbackpopup.ResponseLink);
+    }
+
+    protected override void Resized()
```
What a bizzare layout bug.

beck-thompson replied:
oh yes its quite bizzare 😭

I spent many hours trying to figure it out. All this window stuff is actually from julian (I yoinked it) and he had the magical code which fixes it (I had the same issue in my other version though)

I probably should try to figure out the root cause but I'm unsure exactly where I'd start 🤔

perryprog commented on Content.Server/FeedbackSystem/FeedbackCommand.cs:
```diff
@@ -0,0 +1,67 @@
+            return;
+        }
+
+        if (!int.TryParse(args[0], out var entityUidInt))
```
Would specifying by username be better? (I don't know myself.)

perryprog commented on Content.Server/FeedbackSystem/FeedbackCommand.cs:
```diff
@@ -0,0 +1,67 @@
+
+    public override CompletionResult GetCompletion(IConsoleShell shell, string[] args)
+    {
+        if (args.Length == 1)
+        {
+            return CompletionResult.FromHint(Loc.GetString("feedbackpopup-command-hint-playerUid"));
+        }
+
+        if (args.Length == 2)
+        {
+            return CompletionResult.FromHintOptions(_feedback.FeedbackPopupProtoIds, Loc.GetString("feedbackpopup-command-hint-protoId"));
+        }
+        return CompletionResult.Empty;
```
Switch expression good:
```suggestion
        return args.Length switch
        {
            1 => CompletionResult.FromHint(Loc.GetString("feedbackpopup-command-hint-playerUid")),
            2 => CompletionResult.FromHintOptions(_feedback.FeedbackPopupProtoIds,
                Loc.GetString("feedbackpopup-command-hint-protoId")),
            _ => CompletionResult.Empty
        };
```

perryprog commented on Content.Shared/FeedbackSystem/FeedbackPopupPrototype.cs:
```diff
@@ -0,0 +1,45 @@
+/// Prototype that describes the contents of a feedback popup.
+/// </summary>
+[Prototype]
+public sealed partial class  FeedbackPopupPrototype : IPrototype
```
THE HORRRORRRRR

```suggestion
public sealed partial class FeedbackPopupPrototype : IPrototype
```
(jk <3)

perryprog commented on Content.Shared/FeedbackSystem/FeedbackPopupPrototype.cs:
```diff
@@ -0,0 +1,45 @@
+
+    /// <summary>
+    /// List of "paragraphs" that are placed in the middle of the popup. Put any relevant information about what to give
+    /// feedback on here!
```
Could mention that rich text works here too.

perryprog commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.cs:
```diff
@@ -0,0 +1,60 @@
+    /// </summary>
+    private void LoadPrototypes()
+    {
+        FeedbackPopupProtoIds = _proto.EnumeratePrototypes<FeedbackPopupPrototype>()
+            .Select(x => x.ID)
+            .ToList();
+        FeedbackPopupProtoIds.Sort();
```
Can do:
```suggestion
        FeedbackPopupProtoIds = _proto.EnumeratePrototypes<FeedbackPopupPrototype>()
            .Select(x => x.ID)
            .Order()
            .ToList();
```

perryprog commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.cs:
```diff
@@ -0,0 +1,60 @@
+
+    public bool SendPopups(EntityUid? uid, List<ProtoId<FeedbackPopupPrototype>> popupPrototypes)
+    {
+        if (uid == null)
```
I would just make `uid` not nullable if you can't handle it.

perryprog commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Content.Shared.FeedbackSystem;
+
+public sealed partial class SharedFeedbackSystem : EntitySystem
```
Can just be `FeedbackSystem` since it's sealed.

beck-thompson replied:
Sadly the command seems to mess this up, because the namespaces are the same the import gets sussy

perryprog replied:
Ah I see what you mean. What you can do in the command is add `using FeedbackSystemCSharpIsSoSilly = Content.Shared.FeedbackSystem.FeedbackSystem;` I suppose, but it's up to you of course.

perryprog commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.Events.cs:
```diff
@@ -0,0 +1,50 @@
+
+    private List<string> _validOrigins = [];
+
+    public void EventInitialize()
```
```suggestion
    private void EventInitialize()
```

perryprog commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.Events.cs:
```diff
@@ -0,0 +1,50 @@
+        base.Initialize();
+        SubscribeLocalEvent<RoundEndMessageEvent>(OnRoundEnd);
+
+        _confg.OnValueChanged(CCVars.FeedbackValidOrigins, OnFeedbackOriginsUpdated, true);
```
You can use the Subs helper here:
```suggestion
        Subs.CVar(_confg, CCVars.FeedbackValidOrigins, OnFeedbackOriginsUpdated, true);
```
This automatically unsubscres when the entity system gets shut down so it won't leak a subscription.

beck-thompson commented on Content.Client/FeedbackPopup/FeedbackEntry.xaml.cs:
```diff
@@ -0,0 +1,55 @@
+        // link button
+        if (!string.IsNullOrEmpty(_feedbackpopup.ResponseLink))
+        {
+            LinkButton.OnPressed += OnButtonPressed;
```
LOL yep I had it in the constructor but yeah it gets mad for some reason...

beck-thompson commented:
perry I love the review but this pr is truly evil and now you've forced me to fix it... why 😢 😢 😢 😢 😢

beck-thompson commented:
> I'M NOT A MAINT THO YOU CAN JUST IGNORE ME

your a maint in all but writing sorry to say - your word is law

beck-thompson commented on Content.Client/Stylesheets/StyleNano.cs:
```diff
@@ -1807,6 +1806,29 @@ public StyleNano(IResourceCache resCache) : base(resCache)
                     }),
 
+                // Inset background (News manager, notifications)
+                Element<PanelContainer>().Class("InsetBackground")
```
Did a lot of testing and I think I like this one more, the inset doesn't look quite right imo

beck-thompson replied:
>  don't understand how StyleNano works or how to use it in the slightest

Me neither :sob:

beck-thompson commented:
Thanks for the review perry! Let me know what you think

perryprog commented on Content.Client/FeedbackPopup/FeedbackEntry.xaml.cs:
```diff
@@ -0,0 +1,55 @@
+
+    private void OnButtonPressed(BaseButton.ButtonEventArgs args)
+    {
+        if (_feedbackpopup?.ResponseLink != null)
```
Ah goofy; should be fine.

juliangiebel commented:
> Seperation looks good but it doesn't handle long text well. You need to make the button container's size be dominant and its height locked. 👍
> <img alt="image" width="783" height="766" src="https://private-user-images.githubusercontent.com/83650252/476337665-edfd3112-565b-44fc-ae8f-455459cfc513.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU0NDYwMDEsIm5iZiI6MTc1NTQ0NTcwMSwicGF0aCI6Ii84MzY1MDI1Mi80NzYzMzc2NjUtZWRmZDMxMTItNTY1Yi00NGZjLWFlOGYtNDU1NDU5Y2ZjNTEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODE3VDE1NDgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0OGRhYjExZGZkYmFmMWE4ODMwOGIzNDhlMTg0YTgwOTRhNTYxYjgzNTFmOTJmOWJmMGNjMjhkOWIyZWZkMjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._6A_mBvpc9yG9bvpSrRaD5l8sHhCdURA8lQOo9LRjlM">

I wonder what happened to the code/example I gave. That had enough separation 🤔 
<img width="550" height="493" alt="grafik" src="https://github.com/user-attachments/assets/588a8160-322d-4191-a2ac-147e5186d5f4" />

perryprog commented on Content.Server/FeedbackSystem/FeedbackCommand.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        if (args.Length != 2)
+        {
+            shell.WriteError(Loc.GetString("feedbackpopup-command-error-wrong-arguments"));
```
One of the few times we can get away with reusing loc strings
```suggestion
            shell.WriteError(Loc.GetString("shell-wrong-arguments-number"));
```

perryprog commented on Content.Server/FeedbackSystem/FeedbackCommand.cs:
```diff
@@ -0,0 +1,66 @@
+
+        if (!int.TryParse(args[0], out var entityUidInt))
+        {
+            shell.WriteError(Loc.GetString("feedbackpopup-command-error-invalid-uid"));
```
```suggestion
            shell.WriteError(Loc.GetString("shell-entity-uid-must-be-number"));
```

perryprog commented on Content.Server/FeedbackSystem/FeedbackCommand.cs:
```diff
@@ -0,0 +1,66 @@
+
+        if (!EntityManager.TryGetEntity(netEnt, out var target))
+        {
+            shell.WriteLine(Loc.GetString("feedbackpopup-command-error-entity-not-found"));
```
```suggestion
            shell.WriteLine(Loc.GetString("shell-could-not-find-entity-with-uid", ("uid", entityUidInt)));
```

Rest of the loc strings I think are good. (I'm sus on the existing invalid prototype one.)

perryprog reviewed (commented):
Some other super minor stuff:

SlamBamActionman commented:
I have done some minor changes that I believe are improvements overall:

<img width="698" height="614" alt="image" src="https://github.com/user-attachments/assets/a0519bf2-7fa9-4086-bf50-50229a5596bb" />

1. Some additional padding for the title
2. Made the response type customizable, and changed the count in the bottom right to say "entries" rather than "surveys". 
3. Added the `showfeedbackpopup` command, which shows the available feedback popups. This command is available to all users.

juliangiebel reviewed (approved):
This looks good. It definitly needs a documentation page in the docs.
(It would be better if the commands were toolshed commands though)

Simyon264 commented on Content.Server/FeedbackSystem/FeedbackGivePopupCommand.cs:
```diff
@@ -0,0 +1,69 @@
+/// Give a session a specific feedback pop-up.
+/// </summary>
+[AdminCommand(AdminFlags.Server)]
+public sealed class FeedbackGivePopupCommand : LocalizedEntityCommands
```
This should be a toolshed command in order to minimize boilerplate

Simyon264 commented on Content.Server/FeedbackSystem/FeedbackShowPopupCommand.cs:
```diff
@@ -0,0 +1,26 @@
+/// Show the feedback popups for your own client, if there are any.
+/// </summary>
+[AnyCommand]
+public sealed class FeedbackShowPopupCommand : LocalizedEntityCommands
```
Why is this a command? Wouldn't a net-message be better for this? Or some sort of manager?

juliangiebel replied:
> Why is this a command? Wouldn't a net-message be better for this? Or some sort of manager?

It is a net message. The command just lets you open the pop-up again afaik 🤔

Simyon264 replied:
Ideally you would have this be on something like a keybind or menu button in the action bar (the one where the ahelp or sandbox button live).

beck-thompson replied:
I'm worried about adding more keybinds or buttons on the action bar. We already have so many and even the bug report button got pushback for being too intrusive (and that was much more important than this imo)

Simyon264 commented on Content.Shared/FeedbackSystem/SharedFeedbackSystem.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace Content.Shared.FeedbackSystem;
+
+public sealed partial class SharedFeedbackSystem : EntitySystem
```
This should be out of sim and not a entity system

beck-thompson commented:
I kinda feel like at this point the PR isn't really mine, I already took the UI from Julian and now both Julian and Slam have made changes. Its starting to feel like the bug report stuff all over again and I don't want to repeat that. Of course anyone feel free to make a new PR if you have the motivation to finish this!

juliangiebel commented:
Reopening this from someone elses branch kills the review history. I'm reopening this as someone just needs to move what this adds to stylenano to stylesheets. This isn't something I expected beck to have to do in the first place. Also the reason this has been sitting is the stylenano refactor

juliangiebel commented:
Closed at request of author. A new PR from my branch is here: #41352
